pkg/csi/service/wcp/controller.go

```diff
@@ -1013,10 +1019,77 @@ func (c *controller) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequ
 				"failed to delete volume: %q. Error: %+v", req.VolumeId, err)
 		}
 		if isPodVMOnStretchSupervisorFSSEnabled {
-			err = volumeInfoService.DeleteVolumeInfo(ctx, req.VolumeId)
-			if err != nil {
+			// Fetch the CNSVolumeInfo instance for VolumeID
+			volumeInfo, err := volumeInfoService.GetVolumeInfoForVolumeID(ctx, req.VolumeId)
+			switch {
+			case err == nil:
+				// Get K8s client for StoragePolicyUsage CR.
+				restConfig, err := config.GetConfig()
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to fetch kubernetes config. Error: %+v", err)
+				}
+				cnsOperatorClient, err := k8s.NewClientForGroup(ctx, restConfig, cnsoperatorv1alpha1.GroupName)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to create CNSOperator client. Error: %+v", err)
+				}
+
+				// Fetch StoragePolicyUsage instance for storageClass associated with the volume.
+				storagePolicyUsageInstanceName := volumeInfo.Spec.StorageClassName + "-" +
+					storagepolicyusagev1alpha1.NameSuffixForPVC
+				storagePolicyUsageInstance := &storagepolicyusagev1alpha1.StoragePolicyUsage{}
+				err = cnsOperatorClient.Get(ctx, apitypes.NamespacedName{
+					Namespace: volumeInfo.Spec.Namespace,
+					Name:      storagePolicyUsageInstanceName},
+					storagePolicyUsageInstance)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to fetch %s instance with name %q from supervisor namespace %q. Error: %+v",
+						storagepolicyusagev1alpha1.CRDSingular, storagePolicyUsageInstanceName,
+						volumeInfo.Spec.Namespace, err)
+				}
+
+				// Decrease the used capacity in StoragePolicyUsage instance as we are deleting the volume.
+				newUsedVal := storagePolicyUsageInstance.Status.ResourceTypeLevelQuotaUsage.Used.DeepCopy()
+				newUsedVal.Sub(*volumeInfo.Spec.Capacity)
+
+				// Patch the StoragePolicyUsage instance.
+				patch := map[string]interface{}{
+					"status": map[string]interface{}{
+						"quotaUsage": map[string]interface{}{
+							"used": newUsedVal,
+						},
+					},
+				}
+				patchBytes, err := json.Marshal(patch)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to create patch for StoragePolicyUsage instance. Error: %+v", err)
+				}
+				err = cnsOperatorClient.Patch(ctx, storagePolicyUsageInstance,
```

Comment:
can this be patched multiple time for the same delete operation called again by CSI provisioner?

Reply:
As discussed, we will be setting the owner ref of the PV on the CNSVolumeInfo CR and removing the code to delete the CNSVolumeInfo instance here. With this change, patching the StoragePolicyUsage CR will be the last operation we do before sending out a successful response. If the patch succeeds, there won't be any retry.

Reply:
We can not get PV Owner Ref in the CreateVolume call. We will just get name of the PV which is not sufficient to set ownerRef.
Can we keep this code as it is for now?

Reply:
On second thoughts, this issue needs to investigated further as we cannot set owner reference of PV during in-flight CreateVolume call.